### PR TITLE
docs(samples): Samples for custom vm types

### PR DESCRIPTION
**Pull request by m-strzelczyk**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)

------------------
Some commentary about what is in this code sample. Since the [documentation for custom machine types]() has very little in the means of examples for the API category, I figured it'd be nice to provide our users with something more useful, than just a bunch of strings showing the proper custom machine type string formatting. 

I decided to implement a class, that customers can use to make their life easier when dealing with custom machine types. This class should take care of validating the proper CPU/memory values, provide the "extra memory" switch and handle the formatting of the string for them.

The examples we will then provide in the documentation page will show how to use this class, but also how to use simple strings, if someone doesn't want to use the helper class.

### CODE REVIEW

File: samples/snippets/sample_custom_types.py [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+    """
+    Create a new VM instance with a custom type using shared CPUs.
+
+    Args:
+        project_id: project ID or project number of the Cloud project you want to use.
+        zone: name of the zone to create the instance in. For example: "us-west3-b"
+        instance_name: name of the new virtual machine (VM) instance.
+        cpu_series: The type of CPU you want to use. Pick one value from the CustomMachineType.CPUSeries enum.
+            For example: CustomMachineType.CPUSeries.E2_MICRO
+        memory: Number of megabytes of memory you want to use.
```
Comment (savijatv) [inline samples/snippets/sample_custom_types.py:341]: memory: the amount of memory for the VM instance, in megabytes.
Please change this globally.
It'll be good to provide an example here. For example, 3072MB. Is this correct?

File: samples/snippets/sample_custom_types.py [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+    memory: int,
+):
+    """
+    Create a new VM instance with a custom type using shared CPUs.
+
+    Args:
+        project_id: project ID or project number of the Cloud project you want to use.
+        zone: name of the zone to create the instance in. For example: "us-west3-b"
+        instance_name: name of the new virtual machine (VM) instance.
+        cpu_series: The type of CPU you want to use. Pick one value from the CustomMachineType.CPUSeries enum.
```
Comment (savijatv) [inline samples/snippets/sample_custom_types.py:339]: nit: the type of .. 
Use lower case to make the list consistent.

File: samples/snippets/sample_custom_types.py [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+    project_id: str, zone: str, instance_name: str, core_count: int, memory: int
+):
+    """
+    Create new VM instances without using a CustomMachineType helper function.
+
+    Args:
+        project_id: project ID or project number of the Cloud project you want to use.
+        zone: name of the zone to create the instance in. For example: "us-west3-b"
+        instance_name: name of the new virtual machine (VM) instance.
+        core_count: Number of CPU cores you want to use.
```
Comment (savijatv) [inline samples/snippets/sample_custom_types.py:369]: number of...

File: samples/snippets/sample_custom_types.py [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+# [END compute_custom_machine_type_create_without_helper ]
+
+
+# [START compute_custom_machine_type_extra_mem_no_helper ]
+def create_custom_machines_extra_mem_no_helper(
+    project_id: str, zone: str, instance_name: str, core_count: int, memory: int
+):
+    """
+    Create new VM instances without using a CustomMachineType helper function.
+
```
Comment (savijatv) [inline samples/snippets/sample_custom_types.py:432]: How is this function different from the previous one? Is it using extended memory for the VM? If yes, can we rewrite this comment accordingly?

File: samples/snippets/sample_custom_types.py [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+    Args:
+        project_id: project ID or project number of the Cloud project you want to use.
+        zone: name of the zone to create the instance in. For example: "us-west3-b"
+        instance_name: name of the new virtual machine (VM) instance.
+        core_count: Number of CPU cores you want to use.
+        memory: Number of megabytes of memory you want to use.
```
Comment (savijatv) [inline samples/snippets/sample_custom_types.py:438]: same comment as above for the memory arg

File: samples/snippets/test_sample_custom_types.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    cmt = CustomMachineType(INSTANCE_ZONE, CustomMachineType.CPUSeries.E2_MEDIUM, 8192)
+    assert str(cmt) == f"zones/{INSTANCE_ZONE}/machineTypes/e2-custom-medium-8192"
+    assert cmt.short_type_str() == "e2-custom-medium-8192"
+
+
+def test_custom_machine_type_bad_memory_256():
+    try:
+        CustomMachineType(INSTANCE_ZONE, CustomMachineType.CPUSeries.N1, 8194, 8)
+    except RuntimeError as err:
+        assert err.args[0] == "Requested memory must be multiple of 256 MB."
```
Comment (savijatv) [inline samples/snippets/test_sample_custom_types.py:115]: Missing article. 
...must be a multiple of..

**Review by savijatv — COMMENTED** [review]
@m-strzelczyk , reviewed changes in the files. Are these files reviewed before the changes were made in this PR?

File: samples/snippets/sample_custom_types.py [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+                    )
+            else:
+                raise RuntimeError(
+                    f"Too much memory requested. Maximum memory allowed for {self.cpu_series.name} is {self.limits.max_mem_per_core} MB per core."
+                )
+
+    def __str__(self) -> str:
+        """
+        Present the custom machine type in a format acceptable by GCP API.
+        """
```
[the remark this answers is not in the document]
Comment (m-strzelczyk) [reply samples/snippets/sample_custom_types.py:139]: I'll change it to "Return the custom machine type in form of a string acceptable by Compute Engine API." - this method doesn't display anything anywhere.

File: samples/snippets/sample_custom_types.py [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+        This can use used to create Instance Templates.
+        """
+        return str(self).rsplit("/", maxsplit=1)[1]
+
+    @classmethod
+    def from_str(cls, machine_type: str):
+        """
+        Parse provided string to recreate a custom machine type. The string must be a valid machine type string, for
+        example coming from the API.
+        """
```
[the remark this answers is not in the document]
Comment (m-strzelczyk) [reply samples/snippets/sample_custom_types.py:164]: Yes, I meant the string returned from the API. You get a whole URL of the custom machine type, and you can use this method to turn it into an object of this class.

I'll try to rephrase this:
Construct a new object from a string. The string needs to be a valid custom machine type like:
 - https://www.googleapis.com/compute/v1/projects/diregapic-mestiv/zones/us-central1-b/machineTypes/e2-custom-4-8192
 - zones/us-central1-b/machineTypes/e2-custom-4-8192
 - e2-custom-4-8192 (in this case, the zone parameter will not be set)

File: samples/snippets/sample_custom_types.py [hunk 9]
```diff
@@ -0,0 +1,525 @@
...
+    project_id: str,
+    zone: str,
+    instance_name: str,
+    machine_type: Union[str, "CustomMachineType"],
+):
+    """
+    Send an instance creation request to the Compute Engine API and wait for it to complete.
+
+    Args:
+        project_id: project ID or project number of the Cloud project you want to use.
```
[the remark this answers is not in the document]
Comment (m-strzelczyk) [reply samples/snippets/sample_custom_types.py:222]: Project ID and project number are both unique for every project and they can both be used here. We used this line for the project_id description in all samples so far, so I guess it's OK.

File: samples/snippets/sample_custom_types.py (hunk 4 above)
Comment (m-strzelczyk) [reply samples/snippets/sample_custom_types.py:432]: Yes, the extended memory is the difference. But also, it is not using the helper class, but creates the custom machine type string manually.

File: samples/snippets/sample_custom_types.py [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+            "allowed_cores",
+            "min_mem_per_core",
+            "max_mem_per_core",
+            "allow_extra_memory",
+            "extra_memory_limit",
+        ],
+    )
+
+    LIMITS = {
+        CPUSeries.E2: TypeLimits(range(2, 33, 2), 512, 8192, False, 0),
```
Comment (engelke) [inline samples/snippets/sample_custom_types.py:59]: This value is an iterator, the others are arrays or, in two cases, sets. Is there a reason for the differences?

File: samples/snippets/sample_custom_types.py [hunk 11]
```diff
@@ -0,0 +1,527 @@
...
+        self.cpu_series = cpu_series
+        self.limits = self.LIMITS[self.cpu_series]
+        self.core_count = 2 if self.is_shared() else core_count
+        self.memory_mb = memory_mb
+        self._check()
+        self.extra_memory_used = self._check_extra_memory()
+
+    def is_shared(self):
+        return self.cpu_series in (
+            self.CPUSeries.E2_SMALL,
```
Comment (engelke) [inline samples/snippets/sample_custom_types.py:91]: These three aren't attributes of self, just CPUSeries.E2_SMALL, etc.

File: samples/snippets/sample_custom_types.py [hunk 12]
```diff
@@ -0,0 +1,527 @@
...
+    }
+
+    def __init__(
+        self, zone: str, cpu_series: CPUSeries, memory_mb: int, core_count: int = 0
+    ):
+        self.zone = zone
+        self.cpu_series = cpu_series
+        self.limits = self.LIMITS[self.cpu_series]
+        self.core_count = 2 if self.is_shared() else core_count
+        self.memory_mb = memory_mb
```
Comment (engelke) [inline samples/snippets/sample_custom_types.py:85]: Perhaps add a blank line before _check()

File: samples/snippets/sample_custom_types.py [hunk 13]
```diff
@@ -0,0 +1,527 @@
...
+         - e2-custom-4-8192 (in this case, the zone parameter will not be set)
+        """
+        zone = None
+        if machine_type.startswith("http"):
+            machine_type = machine_type[machine_type.find("zones/") :]
+
+        if machine_type.startswith("zones/"):
+            _, zone, _, type_desc = machine_type.split("/")
+        else:
+            type_desc = machine_type
```
Comment (engelke) [inline samples/snippets/sample_custom_types.py:174]: do you mean maching_type = type_desc?

File: samples/snippets/sample_custom_types.py [hunk 14]
```diff
@@ -0,0 +1,527 @@
...
+        print("Warning during creation:", operation.warnings, file=sys.stderr)
+    print(f"Instance {instance_name} created.")
+    return instance_client.get(project=project_id, zone=zone, instance=instance.name)
+
+
+# [END compute_custom_machine_type_create ]
+
+
+# [START compute_custom_machine_type_create_with_helper ]
+def create_custom_machine(
```
Comment (engelke) [inline samples/snippets/sample_custom_types.py:290]: Maybe "create_custom_instance" to parallel previous method name?

**Review by engelke — APPROVED** [review]
Needs some fixes. See comments and failing tests. Once passing, LGTM.

File: samples/snippets/sample_custom_types.py (hunk 10 above)
Comment (m-strzelczyk) [reply samples/snippets/sample_custom_types.py:59]: Turned everything into frozensets.

File: samples/snippets/sample_custom_types.py (hunk 11 above)
Comment (m-strzelczyk) [reply samples/snippets/sample_custom_types.py:91]: OK, it's not attribute of `self`, but `self` was easier to write than the whole name of the class `CustomMachineType`. I'll change it to `CustomMachineType` though, so it's nicer.

Just `CPUSeries.E2_SMALL` wouldn't work in this context: `NameError: name 'CPUSeries' is not defined`

File: samples/snippets/sample_custom_types.py (hunk 13 above)
Comment (m-strzelczyk) [reply samples/snippets/sample_custom_types.py:174]: I don't think I understand what you mean. I have refactored the code though, as I don't need to switch from using `machine_type` variable to `type_desc`. I can just keep using `machine_type`.

**Pull request merged**